### ISSUE TITLE
Umlauts not supported in Antenna

### ISSUE THREAD

**Issue by tthuem** [tthuem/FeatureIDE#170]
The following project contains a feature model with umlauts in feature names. Even though they are supported in the feature-model editor of Antenna projects, they seem to cause problems during preprocessing. Even features without umlaut in the name are not correctly commented out or in. For example, line

menuOptions.add("Physics Unit");

with configuration

maxConfig.config

in project

https://github.com/DB-SE/isp2014.hannes.klawuhn/tree/master/RechnerAntenna

As we do probably do not want to change Antenna, it seems reasonable to find out the supported letters of Antenna and then to forbid all other characters in the feature model (with the dedicated extension point).

**Comment by tthuem** [comment]
We probably have a regex when parsing Antenna annotations. The feature model should rely on the same regex then.

**Comment by pinnecke** [comment]
I can reproduce this issue. Inside a very tiny project containing a model with two features "base" and its optional subfeature "A" and two configutations where config 1 enables A and config 2 disables A, switching between both configurations works as aspected. By changing "base" to "b**ä**se" the state change for A doesn't work.

**Comment by pinnecke** [comment]
At least 
- umlauts and 
- $ and 
- _
are possbile in feature model but confuses Antenna.

**Comment by pinnecke** [comment]
The *AntennaFMComposerExtension* checks now against a set of illegal characters containing '(', ')', '"', 'ü', 'Ü', 'ä', 'Ä', 'ö', 'Ö', '$', '_'. This restriction is only limited to antenna, which means e.g. the Feature Modeller can contain feature names such as *" Zurück-Button"*. 

Please review this change on remote branch **issue_170** and merge it into master.

**Comment by skrieter** [comment]
Fixed and merged into Master.

**Comment by tthuem** [comment]
The message in the feature model editor "is not a valid Java identifier" is confusing. Those symbols should be named explicitly (as they are indeed valid characters for Java identifiers).

**Comment by pinnecke** [comment]
A description of the set/rules of allowed starting and containing characters for feature names confuses the user more to my mind. What's with a more general message such as:

![bildschirmfoto 2015-02-05 um 15 31 18](https://cloud.githubusercontent.com/assets/3775170/6061933/23f64778-ad4c-11e4-983a-3485ee47014a.png)

?

**Comment by tthuem** [comment]
If there is a long feature name, I would have to check on my own which characters are supported and which are not. That can be frustrating. Hence, please simply print out what characters are not allowed. You may just add them in bracket behind the term "illegal character".

**Comment by pinnecke** [comment]
What do you mean with *check on my own*?

The check for legal feature names is done via built-in java "identifier-character" function. The possible characters set is slight different for starting characters and inner characters. I'm not able to list all **not allowed** characters, because it's {UTF-8}\{Valid_Java_Identifiers}. On the other hand, I can print **allowed** characters, but this is a wall of characters, a set of rules or fuzzy (see oracle description of this function). 

While typing, hitting a **not allowed** character into the textbox shows the user the message above. Hence, he recognizes immediately that his last typed characters is illegal. Of course, copy and past of long feature names will broke this idea, but how often will this happen?

To sum up and make a point: A strait forward solution to inform the users what is allowed and what is not allowed for feature names is achievable with the message above (see common usage of those message, e.g. in several Windows programs). More the less: we block the input and notify the user, that his last character was illegal. From the perspective of the user, that's seems enough to my mind. Printing a huge box with more information, does not inform the user more, that his last character is illegal. If you still want to have a description of what is possible, what's with a **wiki page** for that and why not use a **syntax diagram**?

**Comment by pinnecke** [comment]
Okay, surprisingly the built-in function only takes characters instead of UTF-8. I printed for this set the characters for which both function returned true (= valid)

** Possible characters to start for feature name**
$, A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z, _, a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z

** Possible characters to start for feature name**
$, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z, _, a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z,

**Comment by tthuem** [comment]
OK, what about a message "The feature name is illegal as Antenna does only support $, _, 0-9, a-z, A-Z, whereas 0-9 are not allowed as the first character."?

**Comment by tthuem** [comment]
By the way, isn't that the same as valid Java identifiers? Could also be mentioned and the explanation in brackets only.

**Comment by tthuem** [comment]
I'm not sure whether this is the same or another issue, but the following bug has been introduced in the last days (worked when I gave a demo on January 27th):

![tankwar](https://cloud.githubusercontent.com/assets/5330357/6081429/b7a6cdee-ae19-11e4-8bab-cf767f4bf073.png)

Why should our existing example not work anymore with these feature names? Have you not tried out the Antenna examples when changing the allowed feature names for Antenna?

**Comment by pinnecke** [comment]
If feature model is connected to a composer, the following message will be prompted for *standard* illegal characters (see above). Moreover, for Antenna umlauts are disabled.
![bildschirmfoto 2015-02-10 um 13 54 27](https://cloud.githubusercontent.com/assets/3775170/6127534/5e51bd92-b12c-11e4-8605-db6f789c4ca7.png)

**Comment by tthuem** [comment]
Could you write "Antenna" instead of this composer?

**Comment by pinnecke** [comment]
> By the way, isn't that the same as valid Java identifiers? Could also be mentioned and the explanation in brackets only.

But didn't we removed this information because that's confused the user?

>The message in the feature model editor "is not a valid Java identifier" is confusing. Those symbols should be named explicitly (as they are indeed valid characters for Java identifiers).

We can add this information in brackets, but this isn't necessary to my mind because.

**Comment by pinnecke** [comment]
> Could you write "Antenna" instead of this composer?

Not without more changes in the code itself. That's because the message is a static field in

````
public interface IFMComposerExtension {

	static final String ERROR_MESSAGE_COMPOSER = "The feature name is illegal as this composer does only support $, _, 0-9, a-z, A-Z, whereas 0-9 are not allowed as the first character.";
....
````

and is printed whenever a composer based feature modeller is used.

**Comment by pinnecke** [comment]
> I'm not sure whether this is the same or another issue, but the following bug has been introduced in the last days (worked when I gave a demo on January 27th):

That's because some features in this model uses **underscore** in their names, e.g. Tank**_**Player1 which are forbidden for 13 days (see comment). However, I reviewed it again by deleting **_** from the blacklist and the underscore works in HelloWorld Antenna example. This is surprising, since I figured out that **_** leads to problems which are not reproducible in HelloWorld Antenna example. Unfortunately I cannot remember in which case this problem with **_** occurs. 

Finally, I removed **_** from the blacklist of feature name characters and therefore the tank example works again. But we should have a look at a maybe-problem with **_** in antenna in future.

**Comment by pinnecke** [comment]
# Overview over currently allowed feature names per composer
The following shows the **current state** when FeatureIDE blocks feature names and what are the rules behind this (Analyzed by code reading and testing). 

Given a feature name string ```s```, each feature name have to be **non-null** and **non-empty**. Additionally in context of composer other conditions are checked:
## Feature-Modeller
``s`` is not allowed to contain
*  the **(** character
*  the **)** character
*  the **"** character

## All other composers (general)
```s``` have to pass the ```Character.IsJavaIdentifierStart(char)``` and ```Character.IsJavaIdentifierStart(char)``` test. Depending on the current encoding, **Ü** could be a valid character inside the ```s```.

## Antenna (based on general, but more restricted)
* Additionally ```s``` is not allowed to contain **(**, **)**, **"**, **Ü**, **ü**, **Ä**, **ä**, **ö**, **Ö** and **$**. Maybe **_** could be a problem in the future, but it's currently allowed.

**Comment by pinnecke** [comment]
To inform the user, the following messages will be shown by illegal characters

# Feature-Modeller
![bildschirmfoto 2015-02-10 um 19 41 59](https://cloud.githubusercontent.com/assets/3775170/6133815/ec02338a-b15c-11e4-970e-fed9ed993573.png)

# General all composers
![bildschirmfoto 2015-02-10 um 19 54 19](https://cloud.githubusercontent.com/assets/3775170/6134041/a028b1c6-b15e-11e4-807f-f77371fe19aa.png)

# Antenna
![bildschirmfoto 2015-02-10 um 19 56 18](https://cloud.githubusercontent.com/assets/3775170/6134073/e87edef0-b15e-11e4-8da7-a56456c1e1da.png)

Please note, that this description is based on the Oracle documentation of ```Character.IsJavaIdentifierStart(char)``` and ```Character.IsJavaIdentifierStart(char)```. Of course *Ü* is a valid character inside a feature name but this seems to depend on the charset, more or the less. However this explanation is the most correct and understandable one, to my mind.

**Comment by pinnecke** [comment]
This is available in remote branch **issue_170**, please have a look at.

**Comment by tthuem** [comment]
The branch is 9 commits behind the master, but most probably these commits should not interfere.

The diff seems reasonable and does not indicate any problems to me.

Again, got a compile error because of different encodings

``` Java
if (!Character.isJavaIdentifierPart('Ã¼')) //featureName.charAt(i)))
```
 in FMComposerExtension. Please fix this that I can also compile and run it (i.e., use unicode character encoding).

**Comment by pinnecke** [comment]
Sorry, I fixed it and pushed it to remote branch **issue 170**

**Comment by tthuem** [comment]
I have some remarks to the message box, which we should discuss in a personal meeting. Two issues already now:

1. The umlauts in the message box are not properly shown:

![encoding-bug](https://cloud.githubusercontent.com/assets/5330357/6153854/d0da99be-b227-11e4-9d05-5cc5e3946637.png)

2. The message is also shown if the name is empty, but it does not say that it has to be non-empty.

Everything else worked as expected. You may merge if these two issues are fixed and we talked about the message.

**Comment by tthuem** [comment]
What is the current progress?

**Comment by pinnecke** [comment]
As we talked, I created the following regex for valid feature names:

* in Feature Modelling:  ```.+```
* in Non-Feature Modelling (e.g. Antenna): ```^[a-zA-Z]+\\w*$```

Before testing this above, the feature name is tested against *null*, *empty* and that it does not contain *(*, *)* and *"*. Currently, I look through all examples in feature idea example set to check, if this does not restrict backwards compatibility.

**Comment by tthuem** [comment]
Seems that underscore is missing, right?

**Comment by pinnecke** [comment]
No, underscore is possible because it's in ```.``` set (for feature modeling) and in ```\\w``` (for non feature modeling). Further, the symbols ```[``` and ```]```` are only possible for feature modeling.

**Comment by pinnecke** [comment]
Unfortunately, there is something wrong when reloading the model. 

**Example**: The following model is valid for all composers.
![bildschirmfoto 2015-02-27 um 20 12 27](https://cloud.githubusercontent.com/assets/3775170/6419444/70f2e22a-bebd-11e4-8852-80710b6991ae.png)

Because it's a **feature modeling** model, you can change *B2* to *B [2]*.

![bildschirmfoto 2015-02-27 um 20 12 43](https://cloud.githubusercontent.com/assets/3775170/6419448/7b1a18ea-bebd-11e4-9f0e-6cb1eb9e1472.png)

Which is okay and evaluation works as aspected. You can also switch to editor view

![bildschirmfoto 2015-02-27 um 20 12 54](https://cloud.githubusercontent.com/assets/3775170/6419452/812198bc-bebd-11e4-931a-6ce6058ecb2a.png)

But when **saving**, a re-evaluation starts probably with a wrong `IFMComposerExtension` implementation which leads to an error:

![bildschirmfoto 2015-02-27 um 20 17 50](https://cloud.githubusercontent.com/assets/3775170/6419492/fa7e8b84-bebd-11e4-87ee-fd09768a3230.png)

This is produced by 
```java
if (!featureModel.getFMComposerExtension().isValidFeatureName(name)) {
				throwError(name + " is no valid feature name", e);
			}
``` 

with stack trace:

`
XmlFeatureModelReader.parseFeatures(NodeList, Feature) line: 194	
XmlFeatureModelReader.parseFeatures(NodeList, Feature) line: 222	
XmlFeatureModelReader.parseStruct(NodeList) line: 138	
XmlFeatureModelReader.parseInputStream(InputStream) line: 85	
XmlFeatureModelReader(AbstractFeatureModelReader).readFromString(String) line: 113	
FeatureModelEditor.readModel(AbstractFeatureModelReader, String) line: 636	
FeatureModelEditor.checkModel(String) line: 130	
FeatureModelTextEditorPage.doSave(IProgressMonitor) line: 109	
FeatureModelEditor.doSave(IProgressMonitor) line: 154	
SaveableHelper$2.run(IProgressMonitor) line: 154	
SaveableHelper$5.run(IProgressMonitor) line: 283	
...
`

I'm currently finding out, whats wrong here.

**Comment by pinnecke** [comment]
PS: Changing *B2* to *B [2]* should work as the checking for feature modeling **isValidFeatureName** correctly checks it.

**Comment by tthuem** [comment]
In the tooltip, rather show a-zA-Z_0-9 instead of \\w. That is even accessible to non-programmers.

**Comment by pinnecke** [comment]
## Conclusion

* In **Feature Modeling** every feature name is possible as long as it's non-null and not empty and does not contain *(*,*)* and *"*.
* In **any other composer**: every feature name is possible as long as it's non-null and not empty and does not *(*,*)* and *"* and matches `^[a-zA-Z]+\\w*$`

Further there are not compatibility problems with feature idea built-in examples.

The following warning is shown for illegal feature names in **any composer**:
![bildschirmfoto 2015-02-28 um 17 24 40](https://cloud.githubusercontent.com/assets/3775170/6426876/ea6f6482-bf6e-11e4-92f5-d5362e8e1caf.png)

The following warning is shown for illegal feature names in **feature modeling**:
![bildschirmfoto 2015-02-28 um 17 33 30](https://cloud.githubusercontent.com/assets/3775170/6426895/dbff3d04-bf6f-11e4-88dd-389245f1c7a0.png)

### Next steps
Please review on remote branch **issue_170_mp** and merge it into master. If any action by be are necessary please reassigned this ticket to me.

### Some notes to the current solution
Please note, that I found some restrictions for applying composer specific feature name checks. The current does only work as a **workaround**, that's because the super class ```FMComposerExtension``` checks identically the same for feature names as each composer (inheriting ```FMComposerExtension```, which is not e.g. AHEAD) iff the *hasComposer* flag in ```FMComposerExtension``` is set. However, at least for feature modeling there seems something like a context loss when reloading a model such that ```FMComposerExtension``` checks the name after reloading although ```FeatureModelingComposerExtension``` should do this. If disabling the "copy and paste" name check for *hasComposer* in ```FMComposerExtension``` and delegate it only to sub classes, at least *AspectJ* does strictly use  ```FMComposerExtension``` check instead of it's own. To my understanding these are some problems implied by the current class hierarchy and usage which makes it may hard to restrict per-composer based.  Also at least for feature modeling there is some context loss as described above, which makes this difficult and eventually non-consistent. However, this should be analyzed more in depth in an other dedicated issue if needed. This is not a problem for now, because each composer expect feature modeling is restricted the same way, but I could be a problem in future.

**Comment by tthuem** [comment]
Merged
